scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala

```diff
@@ -238,30 +238,31 @@ private[dynamictls] object DynamicTLSPeerGroupInternals {
                   s"Ssl Handshake server channel from $localAddress " +
                     s"to $remoteAddress with channel id ${ctx.channel().id} and ssl status ${e.isSuccess}"
                 )
-                val channel = new ServerChannelImpl[M](nettyChannel, peerId, codec, messageSubject)
-                serverSubject.onNext(ChannelCreated(channel))
+                val channel = new ServerChannelImpl[M](nettyChannel, peerId, codec, channelQueeue)
+                pushEventOnNettyScheduler(serverQueue, ChannelCreated(channel))(channelScheduler)
```

Comment:
I'm missing something here: why does every instance of `ServerChannelImpl` share the same `channelQueue` (or `messageSubject`)? Doesn't that mean that they all publish into the same common queue/subject? It doesn't seem to make sense then that every channel then had the `in` as a `ConnectableObservable` if it's all just the same one. In fact I don't understand how the handshake protocol works: if it calls `.connect()` on an instance which is common to all channels.

Reply:
In that light I don't understand how this snippet didn't have the problems with the multiple `refCount` invocations:

https://github.com/input-output-hk/scalanet/blob/develop/scalanet/src/io/iohk/scalanet/peergroup/ReqResponseProtocol.scala#L107

Reply:
Each PeerGroup has a global `Subject` (or here queue) which provides incoming channels. i.e `channel.in` provides incoming messages and channel events, `peergroup.in` provides incoming channels and peergroup events.

Reply:
Yes that's how I understand it as well, however:

In `UDPPeerGroup` every `ChannelImpl` that gets created [gets it own](https://github.com/input-output-hk/scalanet/blob/develop/scalanet/src/io/iohk/scalanet/peergroup/UDPPeerGroup.scala#L156) channel event subject, independent of the others.

By contrast in `DynamicTLSPeerGroup` the equivalent is using a [single instance](https://github.com/input-output-hk/scalanet/blob/develop/scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala#L212) of a channel event subject that gets [passed](https://github.com/input-output-hk/scalanet/blob/develop/scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala#L233) to ever incoming channel, so they all publish into the same subject. I think they should be separate.

Reply:
oh ok i see from where the confusion is coming from. Actually this class should probably be called `ServerChannelPiplieBuilder` as its created for each separate incoming channel:
https://github.com/input-output-hk/scalanet/blob/develop/scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroup.scala#L72

So each channel end up with it own builder and its own `messageSubject`

Reply:
Ah, thanks, I knew I was missing something :see_no_evil:

Reply:
By the way, if you consider renaming that class, I'd actually turn it into a method, rather than a constructor that is used as a method, with the resulting class immediately discarded, so this:
```scala
      override def initChannel(ch: SocketChannel): Unit = {
        new ServerChannelBuilder[M](serverSubject, ch, sslServerCtx, codec.cleanSlate)
        log.info(s"$processAddress received inbound from ${ch.remoteAddress()}.")
      }
```
would be 
```scala
     override def initChannel(ch: SocketChannel): Unit = {
        ServerChannelPipelineBuilder.build[M](serverSubject, ch, sslServerCtx, codec.cleanSlate)
        log.info(s"$processAddress received inbound from ${ch.remoteAddress()}.")
      }
```
When I first saw it I wasn't sure the instance wasn't accidentally lost.